src/category_theory/abelian/basic.lean

```diff
@@ -81,20 +81,24 @@ variables (C)
 section prio
 set_option default_priority 100
 
-/-- A (preadditive) category `C` is called abelian if it has a zero object, all binary products and
-    coproducts, all kernels and cokernels, and if every monomorphism is the kernel of some morphism
+/-- A (preadditive) category `C` is called abelian if it has all finite products,
+    all kernels and cokernels, and if every monomorphism is the kernel of some morphism
     and every epimorphism is the cokernel of some morphism. -/
 class abelian extends preadditive.{v} C :=
-[has_zero_object : has_zero_object.{v} C]
-[has_binary_products : has_binary_products.{v} C]
+[has_finite_products : has_finite_products.{v} C]
 [has_kernels : has_kernels.{v} C]
 [has_cokernels : has_cokernels.{v} C]
 (normal_mono : Π {X Y : C} (f : X ⟶ Y) [mono f], normal_mono.{v} f)
 (normal_epi : Π {X Y : C} (f : X ⟶ Y) [epi f], normal_epi.{v} f)
 
-attribute [instance] abelian.has_zero_object
-attribute [instance] abelian.has_binary_products
+attribute [instance] abelian.has_finite_products
 attribute [instance] abelian.has_kernels abelian.has_cokernels
+
+-- Verify that we can find the `has_zero_object C` instance:
+--   `has_finite_products C` gives `has_terminal C`,
+--   and as `preadditive C` gives `has_zero_morphisms C`, together we have `has_zero_object C`.
+example [abelian.{v} C] : has_zero_object.{v} C := by apply_instance
```

Comment:
This seems like a unit test. Why not put this in `tests` in a file that imports just this one?

Reply:
Should I put it in `tests/category_theory/abelian.lean`? or `tests/category_theory.lean`?

Reply:
I'm happy enough to go along with this, but personally I think it's not the right choice. `tests/` is great for unit testing of executable code (e.g. tactics), but to me this line is just another assertion of a mathematical fact.

It's an `example` because anyone wanting the value should just use `by apply_instance` directly.

Besides that though, this feels exactly like a remark in a textbook:

```
\begin{rem}
Notice that our axioms for an abelian category ensure that it has a zero object: finite products imply the existence of a terminal object, and in any preadditive category (indeed, in any category with zero morphisms) a terminal object is zero.
\end{rem}
```

Reply:
I agree this isn't the right choice, and I think this example (or a comment with essentially the remark above) makes it easier to see how to use this file

Reply:
I guess I just don't see what the example is doing then. Either it's testing that this inference problem actually succeeds, in which case it's a unit test; or it's stating for the reader that the hypothesis `abelian C` is enough to derive `has_zero_object C`, in which case there's no need to waste compilation time with an example, and the comment should go somewhere it will actually appear in the docs. 

"Read the source code" isn't a sustainable way to learn a library. I agree that it's the best way to learn mathlib right now. But I think we should be moving away from that, not doubling down.

Reply:
Ok. I've deleted the example, replacing it with a few lines in the doc-string on `abelian`.